<a href="https://colab.research.google.com/github/giridharmickey/data-wrangling/blob/main/column%20drop.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files

df = files.upload()

Saving nss15.csv to nss15.csv


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
data = pd.read_csv('nss15.csv')

In [4]:
#data.head()
#data.describe()
#data.info()
data.columns


Index(['caseNumber', 'treatmentDate', 'statWeight', 'stratum', 'age', 'sex',
       'race', 'diagnosis', 'bodyPart', 'disposition', 'location', 'product'],
      dtype='object')

In [5]:
data.isnull().sum().sum()

129827

In [6]:
data.shape

(334839, 12)

Steps for working with missing data:

-> identify missing data

-> deal with missing data

-> correct data format

**Identify and handle missing values**

In [7]:
data.replace("?", np.nan, inplace = True)
data.head(5)

caseNumber treatmentDate  statWeight stratum  age     sex   race  \
0   150733174     7/11/2015     15.7762       V    5    Male    NaN   
1   150734723      7/6/2015     83.2157       S   36    Male  White   
2   150817487      8/2/2015     74.8813       L   20  Female    NaN   
3   150717776     6/26/2015     15.7762       V   61    Male    NaN   
4   150721694      7/4/2015     74.8813       L   88  Female  Other   

   diagnosis  bodyPart  disposition  location  product  
0         57        33            1         9     1267  
1         57        34            1         1     1439  
2         71        94            1         0     3274  
3         71        35            1         0      611  
4         62        75            1         0     1893

**Evaluating for Missing Data**


The missing values are converted to Python's default. We use Python's built-in functions to identify these missing values. There are two methods to detect missing data:

--> .isnull()

--> .notnull()

In [9]:
missing_data = data.isnull()
missing_data.head(5) #starting 5 rows missing_data = data.isnull().tail() shows last 5 rows

caseNumber  treatmentDate  statWeight  stratum    age    sex   race  \
0       False          False       False    False  False  False   True   
1       False          False       False    False  False  False  False   
2       False          False       False    False  False  False   True   
3       False          False       False    False  False  False   True   
4       False          False       False    False  False  False  False   

   diagnosis  bodyPart  disposition  location  product  
0      False     False        False     False    False  
1      False     False        False     False    False  
2      False     False        False     False    False  
3      False     False        False     False    False  
4      False     False        False     False    False

**"True"** stands for missing value, while **"False"** stands for not missing value.


**Count missing values in each column**

Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value, "False" means the value is present in the dataset. In the body of the for loop the method ".value_counts()" counts the number of "True" values.

In [10]:
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")

caseNumber
False    334839
Name: caseNumber, dtype: int64

treatmentDate
False    334839
Name: treatmentDate, dtype: int64

statWeight
False    334839
Name: statWeight, dtype: int64

stratum
False    334839
Name: stratum, dtype: int64

age
False    334839
Name: age, dtype: int64

sex
False    334837
True          2
Name: sex, dtype: int64

race
False    205014
True     129825
Name: race, dtype: int64

diagnosis
False    334839
Name: diagnosis, dtype: int64

bodyPart
False    334839
Name: bodyPart, dtype: int64

disposition
False    334839
Name: disposition, dtype: int64

location
False    334839
Name: location, dtype: int64

product
False    334839
Name: product, dtype: int64



Based on the summary above, each column and rows of data, one columns containing missing data:

"race": 129825 missing data

**Deal with missing data**

How to deal with missing data?

**1.drop data**

a. drop the whole row

b. drop the whole column

**2.replace data**

a. replace it by mean

b. replace it by frequency

c. replace it based on other functions

**Replace by mean:**

"race": 129825 missing data, replace them with mean

**Replace by frequency:**

"race": 129825 missing data, replace them with "four".

Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur

**Drop the whole row:**

"race": 129825 missing data, simply delete the whole row

Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us

**Calculate the average of the column**

In [11]:
#avg_race = pd.to_numeric(data["race"], errors='coerce').mean()

In [12]:
# Replace null values in each column with the respective mean value
#data["race"].replace(np.nan, avg_race, inplace=True)

#data_filled = data.fillna(data.mean()) #fillna method

In [13]:
data
#data.value_counts() #To see which values are present in a particular column, we can use the ".value_counts()" method
#data.value_counts().idxmax() #We can see that four doors are the most common type. We can also use the ".idxmax()" method
                              #to calculate for us the most common type automatically:

caseNumber treatmentDate  statWeight stratum  age     sex   race  \
0        150733174     7/11/2015     15.7762       V    5    Male    NaN   
1        150734723      7/6/2015     83.2157       S   36    Male  White   
2        150817487      8/2/2015     74.8813       L   20  Female    NaN   
3        150717776     6/26/2015     15.7762       V   61    Male    NaN   
4        150721694      7/4/2015     74.8813       L   88  Female  Other   
...            ...           ...         ...     ...  ...     ...    ...   
334834   150739278     5/31/2015     15.0591       V    7    Male    NaN   
334835   150733393     7/11/2015      5.6748       C    3  Female  Black   
334836   150819286     7/24/2015     15.7762       V   38    Male    NaN   
334837   150823002      8/8/2015     97.9239       M   38  Female  White   
334838   150723074     6/20/2015     49.2646       M    5  Female  White   

        diagnosis  bodyPart  disposition  location  product  
0              57        33            1         9     1267  
1              57        34            1         1     1439  
2              71        94            1         0     3274  
3              71        35            1         0      611  
4              62        75            1         0     1893  
...           ...       ...          ...       ...      ...  
334834         59        76            1         1     1864  
334835         68        85            1         0     1931  
334836         71        79            1         0     3250  
334837         59        82            1         1      464  
334838         57        34            1         9     3273  

[334839 rows x 12 columns]

In [31]:
data.drop('race', axis=1, inplace=True)


**Correct data format**

The last step in data cleaning is checking and making sure that all data is in the correct format (int, float, text or other).


**In Pandas, we use**

--> **.dtype()** to check the data type

--> **.astype()** to change the data type

In [32]:
data.columns

Index(['caseNumber', 'treatmentDate', 'statWeight', 'stratum', 'age', 'sex',
       'diagnosis', 'bodyPart', 'disposition', 'location', 'product'],
      dtype='object')

In [24]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 205014 entries, 1 to 334838
Data columns (total 11 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   caseNumber     205014 non-null  int64  
 1   treatmentDate  205014 non-null  object 
 2   statWeight     205014 non-null  float64
 3   stratum        205014 non-null  object 
 4   age            205014 non-null  int64  
 5   sex            205014 non-null  object 
 6   diagnosis      205014 non-null  int64  
 7   bodyPart       205014 non-null  int64  
 8   disposition    205014 non-null  int64  
 9   location       205014 non-null  int64  
 10  product        205014 non-null  int64  
dtypes: float64(1), int64(7), object(3)
memory usage: 18.8+ MB


In [55]:
X = data [[ 'statWeight', 'age',  'diagnosis', 'bodyPart', 'disposition', 'location', 'product']]
y = data[[ 'statWeight', 'age', 'diagnosis', 'bodyPart', 'disposition', 'location', 'product']]

In [56]:
from sklearn.model_selection import train_test_split

In [57]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=35)

In [58]:
from sklearn.linear_model import LinearRegression

In [82]:
#model

model = LinearRegression()
model.fit(X_train,y_train)



LinearRegression()

In [67]:
y_pred =model.predict(X_test)
y_pred

array([[ 4.96550000e+00,  2.00000000e+00,  5.90000000e+01, ...,
         1.00000000e+00,  1.00000000e+00,  5.02100000e+03],
       [ 1.57762000e+01,  1.30000000e+01,  5.30000000e+01, ...,
         1.00000000e+00,  9.00000000e+00,  1.21100000e+03],
       [ 4.92646000e+01,  2.20000000e+01,  5.90000000e+01, ...,
         1.00000000e+00,  1.00000000e+00,  4.74000000e+02],
       ...,
       [ 1.65650000e+01,  5.70000000e+01,  5.60000000e+01, ...,
         1.00000000e+00, -1.22528197e-15,  8.47000000e+02],
       [ 5.67480000e+00,  1.70000000e+01,  6.40000000e+01, ...,
         1.00000000e+00,  9.00000000e+00,  5.03000000e+03],
       [ 7.85926000e+01,  4.40000000e+01,  5.30000000e+01, ...,
         1.00000000e+00,  2.33497092e-15,  3.22100000e+03]])

In [77]:
from sklearn.metrics import accuracy_score,r2_score,confusion_matrix,precision_score,recall_score,f1_score,mean_absolute_error,mean_squared_error

#Evaluate model
def evaluate_model(y_true, y_pred):
  confusion_matrix = confusion_matrix(y_true,y_pred)
  accuracy_score = accuracy_score(y_true,y_pred)
  r2_score = r2_score(y_true,y_pred)
  precision_score = precision_score(y_true,y_pred)
  recall_score = recall_score(y_true,y_pred)
  f1_score = f1_score(y_true,y_pred)

In [78]:
print('Confusion matrix:\n', confusion_matrix)
print('Accuracy:\n', accuracy_score)
print('r2_score:\n', r2_score)
print('precision:\n', precision_score)
print('recall:\n', recall_score)
print('f1_score:\n', f1_score)


Confusion matrix:
 <function confusion_matrix at 0x7f25fba21510>
Accuracy:
 <function accuracy_score at 0x7f25fba21630>
r2_score:
 <function r2_score at 0x7f25fba82560>
precision:
 <function precision_score at 0x7f25fba21d80>
recall:
 <function recall_score at 0x7f25fba21e10>
f1_score:
 <function f1_score at 0x7f25fba21990>


In [80]:
print(mean_absolute_error(X_test, y_pred))
print(mean_squared_error(X_test, y_pred))

1.1663389534778854e-13
6.338198071976939e-26
